### ISSUE TITLE
noise contrastive estimation/negative sampling

### ISSUE THREAD

**Issue by nsaphra** [pytorch/pytorch#1362]
There isn't a standard loss function implementing this, even though it's pretty common. I am perfectly willing to implement it myself, if nobody else feels like it. It shouldn't be terribly complicated. I would structure it something like:
```python
class NCELoss(torch.nn.modules.loss._Loss):
    r"""Noise contrastive estimation loss function.
    Args:
        num_classes: int number of classes for the output layer
        num_sampled: int number of samples to extract from noise distribution
        noise_sampler: int -> int function
            Function to generate k class labels according to noise distribution.
            By default, noise will be assumed log uniform unigram.

    Shape:
        - Input: :math:`(N, C)` where `C = num_classes`
        - Target: :math:`(N)` where each value is `0 <= targets[i] <= C-1`
    """
```
The API for tensorflow uses a parameter `subtract_log_q` so that by setting it to `False` you can switch to a negative sampling objective from nce.

Is this something that already exists? Is this something worth implementing?

**Comment by apaszke** [comment]
It doesn't exist and we'd be very happy to accept a PR! I'm not familiar with NCE so I can't assess the API until you send a PR, but it looks reasonable.

**Comment by cigrainger** [comment]
If there's anything I can do to help on this I'd be happy to contribute. Code review, bug fixes, testing, etc.

**Comment by soumith** [comment]
@nsaphra one small suggestion is that you dont need to subclass from `torch.nn.modules.loss._Loss`, subclassing from `nn.Module` should be fine.
Having a NCELoss would be great :)

**Comment by cigrainger** [comment]
Is this helpful? https://github.com/analvikingur/pytorch_NEG_loss

**Comment by nsaphra** [comment]
@cigrainger That's a very specific form of negative loss, and not really what I was hoping to make, since I would love to have a general way of combining my nce loss with a multilayer architecture. I would love some advice about how to effectively test my loss function, though, and I would definitely benefit from a code review.

@soumith Why don't I need to subclass? Do you think that there is a documentation/readability benefit from using a loss specific superclass? I guess I'm not sure what the advantage of using `Module` vs. `_Loss`.

**Comment by jekbradbury** [comment]
The `_Loss` subclass is currently tailored to the set of loss functions that wrap existing C fwd+bwd implementations from the Lua Torch backend library. Since you're most likely implementing in pure PyTorch (unless there are performance issues there that you can't get around), it's probably simpler to just subclass `Module`, whose ordinary behavior is to use the PyTorch implementation you provide in `forward` and rely on autograd for the backward pass.
(I think what you're bringing up is effectively the difference between structural and nominative subtyping, and Python seems to lean towards structural).

**Comment by Stonesjtu** [comment]
@nsaphra I have written an NCELoss module that takes the same parameters as `nn.nllloss`. The only difference is the unnormalized probability without softmax . But actually the NCE module should replace the traditional linear decoder at the output layer right before softmax, only in this way the complexity decreases from **|V|** to **k+1**.
Such a module is stateful because it stores the decoder matrix, and, in my opinion it's beyond a **simple** 'Loss' function.
Can I have an early access to your working code?

**Comment by tonyyuango** [comment]
Hi @Stonesjtu, would you please share your NCELoss module? I plan to use ice loss in my current project. Thanks!

**Comment by demelin** [comment]
Hey there, since there has been no recent activity in this thread, I wanted to share my implementation of the NCE loss criterion: https://github.com/demelin/pyTorch_NCE, which came to be as part of a personal project. It's based primarily on the TensorFlow implementation, with some caveats mentioned in the readme. I tried to make the code easy to read and would be be delighted if it ends up being streamlined / extended / improved by the wonderful pyTorch community :) . Thanks!

**Comment by Stonesjtu** [comment]
@tonyyuango  Pushed my training scripts to github. Check this out.
https://github.com/Stonesjtu/Pytorch-NCE/tree/nce

**Comment by tonyyuango** [comment]
@Stonesjtu Thank you so much!

**Comment by nsaphra** [comment]
Sorry, I was out of office for a while. I still think it would be useful to have this in the standard library, and I don't think that the tensorflow license allows basing the implementation on their code. IANAL, but I do think it's all right to use their api to identify what common loss functions might be worth adding beyond this one. Does anyone know?

**Comment by soumith** [comment]
@nsaphra is @Stonesjtu 's implementation fine with you functionally? We can integrate that or a form of that into the standard library.

**Comment by Stonesjtu** [comment]
@nsaphra  I'll take a look at the NCE_loss API in `tensorflow`.

@soumith is there any design docs for the components of `PyTorch`, or any blogs explaining how `Pytorch` works in a deeper manner?

**Comment by parthaca** [comment]
@tonyyuango Have you timed your NCE implementation?

Before I came across this thread, I had implemented a version [here](https://github.com/parthaca/examples/tree/master/word_language_model)
but it turned out to be slower than CE ([details here](https://discuss.pytorch.org/t/nce-loss-for-large-output-vocabularies/3415)). I haven't figured out if it is because of the index_select() and repeat() operations.

**Comment by parthaca** [comment]
@Stonesjtu Sorry, mis-directed my previous comment. Please read the comment above.

**Comment by parthaca** [comment]
It turned out that `torch.multinomial()` was the culprit. I added an implementation of a more efficient sampling method described [here](https://hips.seas.harvard.edu/blog/2013/03/03/the-alias-method-efficient-sampling-with-many-discrete-outcomes/). This makes NCE a bit faster than CE on the 10K PTB dataset and much faster on larger vocabularies. More details in this [thread](https://discuss.pytorch.org/t/nce-loss-for-large-output-vocabularies/3415).

**Comment by tonyyuango** [comment]
Thanks @parthaca and @Stonesjtu! Your implementations help me a lot! I'm new to both pytorch and deep learning, and I just transfer to python from Java, so I encountered some difficulties when reading your code. It would be appreciated if you could clarify them.

I found @parthaca uses the same parameters (weight and bias) for both encoder and nce decoder ("self.decoder.weight = self.encoder.weight" in model.py). I'm not quite clear about the reason. Does it mean that the input and output spaces must be the same? 

I implemented my nce by following @Stonesjtu 's code as it came out a little earlier, but I found the parameters of the decoder don't change during model fitting. Perhaps it's because the decoder is a part of the loss function and its parameters are not involved in the backward propagation (please correct me if I'm wrong). Thus I put the decoder into the RNN model and then it worked.

**Comment by Stonesjtu** [comment]
Q1: It depends on you wether to tie weights of encoder and decoder.
Q2: You should add the parameters of nce to the optimizer, which means you will optimize both your original model and NCE output module.

**Comment by skrish13** [comment]
Are you guys interested in a simple contrastive loss fn as well? Originally proposed in Hadsell et al 2006. Something like [this](https://gist.github.com/harveyslash/725fcc68df112980328951b3426c0e0b#file-contrastive-loss-py)  from [here](https://hackernoon.com/facial-similarity-with-siamese-networks-in-pytorch-9642aa9db2f7)

**Comment by asoltysik** [comment]
@Stonesjtu are there any plans to integrate your work into PyTorch?

**Comment by Stonesjtu** [comment]
I struggled a lot with the choice of API. Currently I found two implementations on `Tensorflow` and `dpnn`. Do you have any suggestions.

Ref:
NCE on tensorflow: https://www.tensorflow.org/api_docs/python/tf/nn/nce_loss
NCE on dpnn: https://github.com/Element-Research/dpnn#nn.NCEModule

**Comment by asoltysik** [comment]
Not really, I am not experienced enough. However, I would recommend making a PR, as I looked through your repo and it seems to be mostly functional? I guess it would be easier to discuss the API and review the PR when submitted.

**Comment by ddehueck** [comment]
Has there been any progress on this? @Stonesjtu I see you've kept your implementation up-to-date, any plans to submit it as a PR?

**Comment by Stonesjtu** [comment]
@ddehueck I will open a PR this weekend, but I should take long time to discuss the API. Another solution is that I can provide a pip package so you don't have to clone the source code.

**Comment by ddehueck** [comment]
@Stonesjtu That would be great. I think having this built-in to PyTorch would be ideal and I'm happy to help with any API discussion in the PR.

**Comment by Santosh-Gupta** [comment]
Does your NCE Loss have log uniform candidate sampling?

**Comment by Stonesjtu** [comment]
@Santosh-Gupta No, because I found uniform sampling gives better PPL for LM on large scale dataset. Another reason is the time complexity of uniform sampling is hugely reduced by using alias method.

**Comment by Santosh-Gupta** [comment]
Was the similarity and analogy properties of the embeddings tested? I find with Unigram sampling the most common words tend to dominate, though I haven't trained a very large scale dataset.

**Comment by Stonesjtu** [comment]
If you mean the norm of the embedding vector of high-frequency words, yes. I've done some word similarity benchmarks, not much difference between sampled loss and cross-entropy loss.

**Comment by Superhzf** [comment]
Any progress on the PR?

**Comment by marcos0318** [comment]
so are there any update on this issue? do we have official code nce loss in pytorch now?
